src/grpc/infer_handler.cc

```diff
@@ -694,6 +694,16 @@ ModelInferHandler::Process(InferHandler::State* state, bool rpc_ok)
   // Handle notification for cancellation which can be raised
   // asynchronously if detected on the network.
   if (state->IsGrpcContextCancelled()) {
+    if (rpc_ok && (state->step_ == Steps::START) &&
+        (state->context_->step_ != Steps::CANCELLED)) {
```

Comment:
question: what does the second clause here imply? 

!= Steps::Cancelled ?

Reply:
To avoid calling `StartNewRequest` twice, at first we fall into HandleCancellation and go through [this](https://github.com/triton-inference-server/server/blob/main/src/grpc/infer_handler.h#L815-L826) block, which returns `true` for `resume`, so we will go into `if (state->IsGrpcContextCancelled()) ` loop for the second time but this time `state->context_->step_ ` is `CANCELLED`

Reply:
Late to the game, but what is the reasoning of not moving the original "`StartNewRequest()` if at START" to before handling the cancellation? Although I think other code needs to be moved around as well.

Reply:
I am not 100% aware of all underlying processes, meaning `state->step_` and `state->context_->step_` combinations. This change helps to address the bug with known symptoms. Refactoring if the `Process` logic needs proper time and testing IMHO

Reply:
@kthui thoughts? If feasible, this can be done as follow-up and by someone else. Want to make sure if there is room for improvement.

Reply:
Yes, I think there is definitely room for improvement/refactoring, i.e. I think the `if (shutdown) { ... }` could also be moved into the `if (state->step_ == Steps::START) { ... } else ...` block, so all procedures for `Steps::START` would be inside the `if (state->step_ == Steps::START) { ... }` block, but it can be done as a follow-up later.

Reply:
Jira ticket: DLIS-6831